### PR TITLE
Total Abundance and Enrichment Fraction - multi

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

Updated the `Total Abundance` and `Enrichment Fraction` `cached_function`s.

## Affected Issue Numbers

- Partially addresses #394
   - [x] PeakGroups:'Total Abundance'...
   - [x] PeakGroups:'Enrichment Fraction'...

## Code Review Notes

Since the enrichment fraction has to be calculated per labeled element the measured compound has in common with the tracer labeled elements, I changed the return value to a dict keyed on element.  I'm not sure how that's going to play out downstream.  (I'm not sure there exists a different model to move it to where the relationship is 1:1, so I invite discussion on that.)

There are some exception catches that could possibly be converted to tests, because given the model structure (e.g. unique constraints), some should never happen, but my thinking was to make the code robust to change/model-editing.

I added a bunch of `TODO` comments that should all be resolved by the end of issue #394.

I couldn't get anything to work until I blew away the migrations.  Let me know if there's a way for me to restore the migrations and re-roll the commits.

The new methods `tracer_labeled_elements` and `common_labels` mirror those of the `accucor_data_loader`.  They take different inputs (because one's for the database and one's for processing file input.  Let me know if you have suggestions on how to consolidate them.

I added a `finally` block under a series of catches just so I don't have repeated code.  It was awkward doing it that way, having to maintain an "error" variable.  Would it be better to have a single `except` line with all the exception types and conditionally handle them based on the exception type/class?

I'm sure there can be lots of improvements that can be made, and I'm eager to make them, so please rip it apart!

## Checklist

- [ ] All issue requirements satisfied (or no linked issues)
   - [x] PeakGroups:'Total Abundance'...
   - [x] PeakGroups:'Enrichment Fraction'...
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
- [ ] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
   - [x] DataRepo.tests.test_models.PropertyTests.test_peak_group_peak_data_1
   - [x] DataRepo.tests.test_models.PropertyTests.test_peak_group_peak_data_4
   - [x] DataRepo.tests.test_models.PropertyTests.test_peak_group_peak_data_serum
   - [x] DataRepo.tests.test_models.PropertyTests.test_enrichment_fraction_missing_compounds
   - [x] DataRepo.tests.test_models.PropertyTests.test_enrichment_fraction_missing_labeled_element
   - [x] DataRepo.tests.test_models.PropertyTests.test_peak_group_total_abundance_zero
   - [x] DataRepo.tests.test_models.PropertyTests.test_peak_group_common_labels
   - [x] DataRepo.tests.test_models.PropertyTests.test_peak_group_tracer_labeled_elements
   - [x] DataRepo.tests.test_models.PropertyTests.test_no_common_labels

### CODE REVIEW

File: DataRepo/models/peak_group.py [hunk 1]
```diff
@@ -53,50 +53,135 @@ def total_abundance(self):
...
+                if (
+                    compound_rec.atom_count(tracer_labeled_element) > 0
+                    and tracer_labeled_element not in common_labels
+                ):
+                    common_labels.append(tracer_labeled_element)
+        return common_labels
+
+    @property  # type: ignore
+    @cached_function
+    def enrichment_fractions(self):
```
Comment (lparsons) [inline DataRepo/models/peak_group.py:101]: C: It seems to me that this should be something that requires a parameter (the element) and then calculates the values the same way it did before, but using `peak_data.fraction` and `peak_data.labled_count` also with the element as a parameter. We could provide some wrapper that gets all of the elements and then puts that in a dictionary if we want that type of return.

**Review by lparsons — COMMENTED** [review]
This seems to work, but it's a bit hard to follow. I'm thinking the calculations and logic would be simpler to suggest breaking the functions up a bit and providing the element as a parameter.

File: DataRepo/formats/peakgroups_dataformat.py [hunk 2]
```diff
@@ -139,12 +139,14 @@ class PeakGroupsFormat(Format):
                     "displayed": True,
                     "type": "string",
                 },
-                "enrichment_fraction": {
-                    "displayname": "Enrichment Fraction",
-                    "searchable": False,  # Cannot search cached property
-                    "displayed": True,
-                    "type": "number",
-                },
+                # TODO: This property/cached_function returns a dict, a type that is not (yet) supported.  Commenting
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/formats/peakgroups_dataformat.py:142]: Meh. A lot of them are related, and I feel like using the checkboxes in the issue and TODO tags to coordinate work well in this case, because I'm coming across the related code as I work on one component...  I just figured I'd submit PRs at stopping points.

File: DataRepo/models/peak_group.py [hunk 3]
```diff
@@ -53,50 +53,135 @@ def total_abundance(self):
...
+        observations (i.e. count=0 and parent=True).  This is so that Isocorr data canrecord 0 observations for parent
+        records.  Accucor data does present data for counts of 0 already.
+        """
+        # Assuming all samples come from 1 animal, so we're only looking at 1 (any) sample
+        tracer_labeled_elements = []
+        for tracer in self.msrun.sample.animal.infusate.tracers.all():
+            for label in tracer.labels.all():
+                if label.element not in tracer_labeled_elements:
+                    tracer_labeled_elements.append(label.element)
+        return tracer_labeled_elements
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/peak_group.py:78]: 👉🏻👃🏻 Step ahead of you.  I hasd the same thought and moved it in #472 just before you submitted your review here of 471.

File: DataRepo/models/peak_group.py [hunk 4]
```diff
@@ -53,50 +53,135 @@ def total_abundance(self):
...
+
+    @property  # type: ignore
+    @cached_function
+    def common_labels(self):
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/peak_group.py:82]: I'll change it because I agree it's not precise enough, though I would like there to be a distinction between labeled_elements among the tracers versus the subset of labeled elements common between the tracers and the measured compound.

- [ ] Change `common_labels` to `peak_labeled_elements`

File: DataRepo/models/peak_group.py [hunk 5]
```diff
@@ -53,50 +53,135 @@ def total_abundance(self):
...
+        """
+        Gets labels present among any of the tracers in the infusate IF the elements are present in the supplied
+        (measured) compounds.  Basically, if the supplied compound contains an element that is a labeled element in any
+        of the tracers, included in the returned list.
+        """
+        common_labels = []
+        compound_recs = self.compounds.all()
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/peak_group.py:89]: Sounds like that would more efficient.  We can create a cached_function or maintained field that saves a dict with all the atom counts for each compound.  Not sure that there would be an observable difference, so I'll just create an issue... #474

File: DataRepo/models/peak_group.py (hunk 1 above)
Comment (hepcat72) [reply DataRepo/models/peak_group.py:101]: I've been thinking about moving this into PeakDataLabel where no parameter is needed because I think that caching or field maintenance would be more complicated to handle a parameter.  I will do this in #472

File: DataRepo/models/peak_group.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/peak_group.py:82]: I'll do this in #472

File: DataRepo/models/peak_group.py [hunk 6]
```diff
@@ -53,50 +53,135 @@ def total_abundance(self):
...
+                    element_enrichment_sum = element_enrichment_sum + (
+                        label_pd_rec.fraction * label_rec.count
+                    )
+
+                # Calculate the denominator
+                # This assumes that multiple measured compounds for the same PeakGroup are composed of the same elements
+                atom_count = compound.atom_count(measured_element)
+
+                enrichment_fractions[measured_element] = (
+                    element_enrichment_sum / atom_count
```
Comment (jcmatese) [inline DataRepo/models/peak_group.py:144]: Q: I don't suppose `atom_count` could ever be 0?

**Review by jcmatese — APPROVED** [review]
Sounds like some of this was moved to another PR, so I don't think I should hold this up, given incremental progress towards the eventual goal.

**Review by jcmatese — COMMENTED** [review]
Sorry, did not see that this was just Draft/WIP

File: DataRepo/models/peak_group.py (hunk 6 above)
Comment (hepcat72) [reply DataRepo/models/peak_group.py:144]: This is old code (see old line 83). However, we're looking at elements in this loop that have been cross-referenced with the compound (see the common_labels variable), so if it's ever zero, that would be a problem with the atom_count method and I think we would want an exception if something like that wasn't working IMHO.

**Comment by hepcat72** [comment]
With the caveat that a bit of this code is changed in the next PR, I have taken this out of draft.

**Review by lparsons — APPROVED** [review]
Approving since this didn't seem to change since my last review, it appears the changes are in #476 instead.

Some comments:

* I like the idea of creating a table with a row for each PeakGroup.element to house the enrichment_fraction, etc. It seems cleaner than returning a `dict`.

* > I couldn't get anything to work until I blew away the migrations. Let me know if there's a way for me to restore the migrations and re-roll the commits.
  
  Without seeing the exact errors I don't think I can help with this. That said, I'm ok recreating the databases at this point for the sake of moving forward.

* Parsing strings vs querying the database seems to be different enough to me that we don't need to consolidate the `PeakGroup.tracer_labeled_elements` with `AccuCorDataLoader.get_tracer_labels`. Same with `PeakGroup.common_labels` and `AccuCorDataLoader.get_common_labels`.

* The `finally` block seems readable enough to me, though it seems like we might want to be raising those exceptions instead of tracing an "error". I do think the logic here could be streamlined, but it seems good enough to progress with.

**Comment by jcmatese** [comment]
I agree on the creation of a hypothetical "peak_group_derived_data" model, which I think we had discussion on, but probably never got around to codifying.  I think I saw an [issue](https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/477) created for that.  I say press on, and we will keep refining.  Sorry I haven't been much assistance.
